helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -125,6 +211,79 @@ public void close() {
     _baseDataAccessor.close();
   }
 
+  @Override
+  public void handleDataChange(String dataPath, Object data) throws Exception {
+    Stat stat = new Stat();
+    ZNRecord readData =
+        _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+    LockInfo lockInfo = new LockInfo(readData);
+    // We are the current owner
+    if (lockInfo.getOwner().equals(_userId)) {
+      if (lockInfo.getRequestorId() == null || lockInfo.getPriority() > lockInfo
+          .getRequestorPriority()) {
+        LOG.info("We do not need to handle this data change");
+      } else {
+        _lockListener.onCleanupNotification();
+        // read the lock information again to avoid stale data
+        readData = _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+        // If we are still the lock owner, clean the lock owner field.
+        if (lockInfo.getOwner().equals(_userId)) {
+          ZNRecord znRecord = new ZNRecord(readData);
+          znRecord.setSimpleField(LockInfo.LockInfoAttribute.OWNER.name(), "");
+          znRecord.setSimpleField(LockInfo.LockInfoAttribute.MESSAGE.name(), "");
+          znRecord.setLongField(LockInfo.LockInfoAttribute.TIMEOUT.name(), -1);
+          znRecord.setIntField(LockInfo.LockInfoAttribute.PRIORITY.name(), -1);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.WAITING_TIMEOUT.name(), -1);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.CLEANUP_TIMEOUT.name(), -1);
+          _baseDataAccessor
+              .update(_lockPath, new ZNRecordUpdater(znRecord), AccessOption.PERSISTENT);
+        } else {
+          LOG.info("We are not current lock owner");
+        }
+      }
+    } // We are the current requestor
+    else if (lockInfo.getRequestorId().equals(_userId)) {
```

Comment:
I cannot match it easily with the if condition above, but they look like the same : )

Reply:
The if condition is:
if (lockInfo.getOwner().equals(_userId))

Basically we need to distinguish between requestor and owner. Both of them may receive callback, but they don't need to handle unrelated callback.